File: csharp/GooseFeatures/SessionManager.cs
// Re-implementation of Goose's Session and Memory Management

using System.Collections.Generic;

namespace TormentNexus.GooseFeatures
{
    public class Session
    {
        public string Id { get; }
        public List<Message> History { get; }

        public Session(string id)
        {
            Id = id;
            History = new List<Message>();
        }

        public void AddMessage(Message message)
        {
            History.Add(message);
        }

        public IReadOnlyList<Message> GetHistory()
        {
            return History.AsReadOnly();
        }
    }

    public class SessionManager
    {
        private readonly Dictionary<string, Session> _sessions = new Dictionary<string, Session>();

        public Session CreateSession(string id)
        {
            var session = new Session(id);
            _sessions[id] = session;
            return session;
        }

        public Session GetSession(string id)
        {
            _sessions.TryGetValue(id, out var session);
            return session;
        }

        public void ClearSession(string id)
        {
            _sessions.Remove(id);
        }
    }
}

File: go/internal/goose/session_manager.go
package goose

// Re-implementation of Goose's Session and Memory Management

type Session struct {
	ID      string
	History []Message
}

func NewSession(id string) *Session {
	return &Session{
		ID:      id,
		History: []Message{},
	}
}

func (s *Session) AddMessage(msg Message) {
	s.History = append(s.History, msg)
}

func (s *Session) GetHistory() []Message {
	return s.History
}

type SessionManager struct {
	sessions map[string]*Session
}

func NewSessionManager() *SessionManager {
	return &SessionManager{
		sessions: make(map[string]*Session),
	}
}

func (sm *SessionManager) CreateSession(id string) *Session {
	session := NewSession(id)
	sm.sessions[id] = session
	return session
}

func (sm *SessionManager) GetSession(id string) (*Session, bool) {
	session, exists := sm.sessions[id]
	return session, exists
}

func (sm *SessionManager) ClearSession(id string) {
	delete(sm.sessions, id)
}
